README

			README for liblockfile


This library implements a number of functions found in -lmail on SysV
systems. These functions are designed to lock the standard mailboxes in
/var/mail (or wherever the system puts them).

In additions, this library adds a number of functions to create,
manage and remove generic lockfiles.

The lockfiles are created by this library in an NFS-safe manner, that
is by using a tempfile and using link(2) to create the lockfile. It
works around several defects in NFS servers to make sure the lockfile
is created atomically.

The locking strategy is compatible with other well-written NFS-safe
programs that create lockfiles - such as procmail, exim and mutt.

If you are trying to lock a mailbox in a directory writable for group
mail, the locking functions will call a helper application which runs
setgid to do the actual locking. The helper application "dotlockfile"
can also be used directly, for example for use in shellscripts.

This means a program such as a MUA doesn't need to be setgid mail anymore
to be able to lock the mailbox.

See the included manualpages for more info:

Function			Manpage
========			=======
maillock,
mailunlock,
touchlock -			maillock.3

lockfile_create,
lockfile_remove,
lockfile_touch,
lockfile_check -		lockfile_create.3




Changelog
liblockfile (1.17)

  * dotlockfile: '-P' option: allow passing through the child exit code
  * dotlockfile: '-i interval' option.
  * dotlockfile: document -q
  * add L_RMSTALE return status (failed to remove stale lock)
  * remove cistron.nl email address

 -- Miquel van Smoorenburg Wed, 27 Jan 2021 00:33:10 +0100

    v1.16

liblockfile (1.16)

  * lockfile.c: remove all the permission checking in the normal case.
    Only if creating the tempfile / removing the lockfile returns EACCESS,
    check to see if this is a mail-lockfile, if so launch helper.
  * dotlockfile.c: use same mail-lockfile check as lockfile.c, but a
    bit more strict.
  * dotlockfile: actually honor -q flag (quiet) everywhere

 -- Miquel van Smoorenburg <miquels@cistron.nl>  Mon, 29 Jul 2019 19:31:09 +0200

liblockfile (1.15)

  * dotlockfile.1: Fix groff syntax (from 1.14-1.1 debian NMU by Mattia Rizzolo <mattia@debian.org>)
  * dotlockfile.1: Some minor fixes in the manual (from https://bugs.debian.org/899344)
  * fix getgroups() check in lockfile.c (Thomas Finke), add some cleanups
  * fix gcc warning about snprintf in lockfile_create_save_tmplock() (not an actual bug)
  * fix gcc warning about strncpy in ismaillock() (not an actual bug)
  * add support for systems that do not have PATH_MAX
    (fixes https://github.com/miquels/liblockfile/issues/1)
  * rename configure.in to configure.ac and run autoconf
    (fixes https://github.com/miquels/liblockfile/issues/2)

 -- Miquel van Smoorenburg <miquels@cistron.nl>  Tue, 23 Jul 2019 23:25:53 +0200

liblockfile (1.14)

  * allow '-m' option again when executing command (why disallow it)
  * drop privs just after fork (and before execv) instead of earlier
  * before executing command, chdir back to original working directory.

 -- Miquel van Smoorenburg <miquels@cistron.nl>  Tue, 17 Jan 2017 13:58:25 +0100

liblockfile (1.13)

  * remove '-t' as option when running a command- do touchlock
    automatically if -p was not given.
  * drop effective gid asap, retain as saved gid. only restore egid if necessary.
  * if stat() fails, drop saved gid as well, and retry (sshfs is strict)
  * don't allow executing a command with the -m option, not necessary.
    should really add a minimal setgid util for maillock(), remove setgid
    from dotlockfile, and let it use the library maillock().

 -- Miquel van Smoorenburg <miquels@cistron.nl>  Thu, 05 Jan 2017 23:10:35 +0100

liblockfile (1.12)

  * update lockfile_create.3 manpage, document L_PPID

 -- Miquel van Smoorenburg <miquels@cistron.nl>  Thu, 05 Jan 2017 11:37:22 +0100

liblockfile (1.11)

  * fix dotlockfile -r -1 functionality
  * add dotlockfile <lockfile> cmd args... functionality
  * dotlockfile: use geteuid() instead of getuid() to find out mailboxname.
  * before forking an external program, first check geteuid() != 0,
    then do setuid(geteuid())

 -- Miquel van Smoorenburg <miquels@cistron.nl>  Tue, 03 Jan 2017 20:45:15 +0100

liblockfile (1.10)

  * add complete GPL/LGPL licenses to source code
  * Merge all debian patches into upstream
  * fail to lock if parent is PID 1 - makes no sense to write PID 1
    as owning process into the lockfile (Per Cederqvist <cederp@opera.com>)

 -- Miquel van Smoorenburg <mike@langeraar.net>  Fri, 18 Nov 2016 14:14:28 +0100

liblockfile (1.09)

  * Update path to full license in the COPYRIGHT file
    (Anibal Monsalve Salazar)
  * Remove debug code from lockfile_create():
     The lockfile_create-function still contains some debug code, which
     considers a lock stale after 10 seconds instead of 5 minutes. As
     the documentation suggests refreshing the lock every 60 seconds,
     programs might lose the lock unexpectedly. Found while researching
     debian bug #505278. (Wolfram Sang <w.sang@pengutronix.de>)
  * Updated dotlockfile manpage and updated dotlockfile usage message.
    Fixes debian bug 493462 (Roland Eggner <edvd1@systemanalysen.net>)
  * Change 'ar' in Makefile.in to '$(AR)' (for cross builds).
    Fixes debian bug 562937 (Wolfram Sang <wsa@pengutronix.de>)
  * Clean up temporary lockfile if lockfile(1) was interrupted.
  * dotlockfile now also allows you to create a lockfile in
    a group-writable-by-one-of-your-auxilary-groups directory
  * dotlockfile: -r-1 now retries indefinitely (well, 4000 years)
  * remove ppsh macro in dotlockfile.1, replace with standard .SH
    (Iain Calder <dn097@torfree.net>)

 -- Miquel van Smoorenburg <miquels@cistron.nl>  Wed, 20 Jul 2011 15:54:58 +0200

liblockfile (1.08)

  * When not writing a pid into the lockfile, write "0\n" as
    placeholder instead of "0\0" (bug reported by Rob Browning
    <rlb@defaultvalue.org>)
  * lockfile_check() did the following:
      - open lockfile
      - read from it
      - stat file and get atime as 'now'
    this gets the 'now' time from the filesystem so comparisations
    to mtime work even if the clock of the remote file server is off.
    However this doesn't work with noatime / relatime, so now we stat
    the lockfile both before and after the read and use atime only
    if it differs before and after.
  * Copy Changelog from debian/changelog now that the package
    isn't native anymore since 1.07


run-tests.sh
#!/bin/sh
#
#	Some simple tests to check if locking/unlocking works correctly.
#

set -e
PATH=.:$PATH

rm -f testlock.lock

# lock should succeed
dotlockfile -l -r 1 testlock.lock

# locking again should fail
! dotlockfile -l -r 0 testlock.lock

# unlocking should succeed
dotlockfile -u testlock.lock

[ ! -f testlock.lock ] || { echo "lockfile still exists after unlock"; exit 1; }

# same but with a command
dotlockfile -l -r 1 testlock.lock sleep 2 &
PID=$!
sleep 1

# locking again should fail
if dotlockfile -l -r 0 testlock.lock; then
    if kill -0 "$PID" 2>/dev/null; then
        echo "double lock while running command"
        exit 1
    fi
fi

# lock should be gone when command completes
wait
[ ! -f testlock.lock ] || { echo "lockfile still exists after running cmd"; exit 1; }

# test locking with given sleep interval
dotlockfile -l -r 0 testlock.lock
dotlockfile -l -i 4 testlock.lock /bin/true &

time_start=$(date '+%s')

sleep 4.5
dotlockfile -u testlock.lock

wait

time_end=$(date '+%s')
time_elapsed=$((time_end - time_start))

# Time should equal 8 seconds
[ "$time_elapsed" = 8 ] || { echo "lockfile should take 8 seconds to be replaced. [$time_elapsed]"; exit 1; }
[ ! -f testlock.lock ] || { echo "lockfile still exists after running cmd"; exit 1; }

echo "tests OK"


lockfile.h
/*
 *	Copyright (C) 1999 Miquel van Smoorenburg
 *
 *	This library is free software; you can redistribute it and/or modify
 *	it under the terms of the GNU Library General Public License as
 *	published by the Free Software Foundation; either version 2 of the
 *	License, or (at your option) any later version.
 *
 *	On Debian GNU/Linux systems, the complete text of the GNU Library
 *	General Public License can be found in `/usr/doc/copyright/LGPL'.
 *	You can also find a copy on the GNU website at http://www.gnu.org/
 */
#ifndef _LOCKFILE_H
#define _LOCKFILE_H

#ifdef  __cplusplus
extern "C" {
#endif

/*
 *	Prototypes.
 */
int	lockfile_create(const char *lockfile, int retries, int flags, ...);
int	lockfile_remove(const char *lockfile);
int	lockfile_touch(const char *lockfile);
int	lockfile_check(const char *lockfile, int flags);

/*
 *	Return values for lockfile_create()
 */
#define	L_SUCCESS	0	/* Lockfile created			*/
#define L_NAMELEN	1	/* Recipient name too long 		*/
#define L_TMPLOCK	2	/* Error creating temp lockfile		*/
#define L_TMPWRITE	3	/* Can't write pid into temp lockfile	*/
#define L_MAXTRYS	4	/* Failed after max. number of attempts	*/
#define L_ERROR		5	/* Unknown error; check errno		*/
#define L_MANLOCK	6	/* Cannot set mandatory lock on tempfile */
#define L_ORPHANED	7	/* Called with L_PPID but parent is gone */
#define L_RMSTALE	8	/* Failed to remove stale lockfile	*/

/*
 *	Flag values for lockfile_create()
 */
#define L_PID		16	/* Put PID in lockfile			*/
#define L_PPID		32	/* Put PPID in lockfile			*/
#define L_INTERVAL	64	/* Specify consistent retry interval	*/

/*
 * Optional arguments for lockfile_create.
 */
struct lockargs {
	int interval;		/* Static interval between retries	*/
};

#ifdef  __cplusplus
}
#endif

#endif /* _LOCKFILE_H */


lockfile.c
/*
 * lockfile.c 	Safely creates a lockfile, also over NFS.
 *		This file also holds the implementation for
 *		the Svr4 maillock functions.
 *
 *		Copyright (C) Miquel van Smoorenburg and contributors 1997-2021.
 *
 *		This library is free software; you can redistribute it and/or
 *		modify it under the terms of the GNU Library General Public
 *		License as published by the Free Software Foundation; either
 *		version 2 of the License, or (at your option) any later version.
 */

#include "autoconf.h"

#include <sys/types.h>
#if HAVE_SYS_PARAM_H
#include <sys/param.h>
#endif
#include <sys/stat.h>
#include <sys/wait.h>
#include <stdarg.h>
#include <stdio.h>
#include <string.h>
#include <signal.h>
#include <fcntl.h>
#include <stdlib.h>
#include <unistd.h>
#include <time.h>
#include <errno.h>
#include <lockfile.h>
#include <maillock.h>

#ifdef HAVE_UTIME
#include <utime.h>
#endif

#ifdef LIB
static char *mlockfile;
static int  islocked = 0;
#endif

#ifndef LIB
extern int check_sleep(int, int);
#endif

#ifdef MAILGROUP
/*
 *	Get the id of the mailgroup, by statting the helper program.
 *	If it is setgroup-id, then the group is the mailgroup.
 */
static int mailgid()
{
	struct stat st;

	if (stat(LOCKPROG, &st) != 0)
		return (gid_t)-1;
	if ((st.st_mode & 02000) == 0)
		return (gid_t)-1;
	return st.st_gid;
}

/*
 *	Is this a lock for a mailbox? Check:
 *	- is the file in /path/to/USERNAME.lock format
 *	- is /path/to/USERNAME present and owned by us
 *	- is /path/to writable by group mail
 *
 *	To be safe in a setgid program, chdir() into the lockfile
 *	directory first, then pass in the basename of the lockfile.
 */
#ifdef LIB
static
#endif
int is_maillock(const char *lockfile)
{
	struct stat	st;
	gid_t		gid;
	char		tmp[1024];
	char		*p;

	/* remove .lock suffix */
	strncpy(tmp, lockfile, sizeof(tmp) - 1);
	tmp[sizeof(tmp) - 1] = 0;
	if ((p = strrchr(tmp, '.')) == NULL || strcmp(p, ".lock") != 0)
		return 0;
	*p = 0;

	/* file to lock must exist, and must be owned by us */
	if (lstat(tmp, &st) != 0 ||
	    (st.st_mode & S_IFMT) != S_IFREG || st.st_uid != getuid())
		return 0;

	/* Directory this file is in must be writable by group mail. */
	if ((gid = mailgid()) == (gid_t)-1)
		return 0;
	if ((p = strrchr(tmp, '/')) != NULL)
		*p = 0;
	else
		strncpy(tmp, ".", sizeof(tmp));
	if (stat(tmp, &st) != 0 || st.st_gid != gid || (st.st_mode & 0020) == 0)
		return 0;

	return 1;
}

#ifdef LIB
/*
 *	Call external program to do the actual locking.
 */
static int run_helper(char *opt, const char *lockfile, int retries, int flags)
{
	sigset_t	set, oldset;
	char		buf[8];
	pid_t		pid, n;
	int		st;

	/*
	 * Better safe than sorry.
	 */
	if (geteuid() == 0)
		return L_ERROR;

	/*
	 *	Block SIGCHLD. The main program might have installed
	 *	handlers we don't want to call.
	 */
	sigemptyset(&set);
	sigaddset(&set, SIGCHLD);
	sigprocmask(SIG_BLOCK, &set, &oldset);

	/*
	 *	Fork, execute locking program and wait.
	 */
	if ((pid = fork()) < 0)
		return L_ERROR;
	if (pid == 0) {
		/* drop privs */
		if (setuid(geteuid()) < 0) {
			perror("setuid");
			_exit(L_ERROR);
		}
		snprintf(buf, sizeof(buf), "%d", retries % 1000);
		execl(LOCKPROG, LOCKPROG, opt, "-r", buf, "-q",
			(flags & L_PID) ? "-p" : "-N", lockfile, NULL);
		_exit(L_ERROR);
	}

	/*
	 *	Wait for return status - do something appropriate
	 *	if program died or returned L_ERROR.
	 */
	while ((n = waitpid(pid, &st, 0)) != pid)
		if (n < 0 && errno != EINTR)
			break;
	if (!sigismember(&oldset, SIGCHLD))
		sigprocmask(SIG_UNBLOCK, &set, NULL);
	if (n < 0)
		return L_ERROR;
	if (!WIFEXITED(st) || WEXITSTATUS(st) == L_ERROR) {
		errno = EINTR;
		return L_ERROR;
	}

	return WEXITSTATUS(st);
}
#endif /* LIB*/

#endif /* MAILGROUP */

#define TMPLOCKSTR		".lk"
#define TMPLOCKSTRSZ		strlen(TMPLOCKSTR)
#define TMPLOCKPIDSZ		5
#define TMPLOCKTIMESZ		1
#define TMPLOCKSYSNAMESZ	23
#define TMPLOCKFILENAMESZ	(TMPLOCKSTRSZ + TMPLOCKPIDSZ + \
				 TMPLOCKTIMESZ + TMPLOCKSYSNAMESZ)

static int lockfilename(const char *lockfile, char *tmplock, int tmplocksz)
{
	char		sysname[256];
	char		*p;

#ifdef MAXPATHLEN
	/*
	 *	Safety measure.
	 */
	if (strlen(lockfile) + TMPLOCKFILENAMESZ > MAXPATHLEN) {
		errno = ENAMETOOLONG;
		return L_ERROR;
	}
#endif

	if (strlen(lockfile) + TMPLOCKFILENAMESZ + 1 > tmplocksz) {
		errno = EINVAL;
		return L_ERROR;
	}

	/*
	 *	Create a temp lockfile (hopefully unique) and write
	 *	either our pid/ppid in it, or 0\0 for svr4 compatibility.
	 */
	if (gethostname(sysname, sizeof(sysname)) < 0)
		return L_ERROR;
	if ((p = strchr(sysname, '.')) != NULL)
		*p = 0;
	/* strcpy is safe: length-check above, limited at snprintf below */
	strcpy(tmplock, lockfile);
	if ((p = strrchr(tmplock, '/')) == NULL)
		p = tmplock;
	else
		p++;
	if (snprintf(p, TMPLOCKFILENAMESZ, "%s%0*d%0*x%s", TMPLOCKSTR,
			TMPLOCKPIDSZ, (int)getpid(),
			TMPLOCKTIMESZ, (int)time(NULL) & 15,
			sysname) < 0) {
		// never happens but gets rid of gcc truncation warning.
		errno = EOVERFLOW;
		return L_ERROR;
	}

	return 0;
}

/*
 *	Create a lockfile.
 */
static int lockfile_create_save_tmplock(const char *lockfile,
		char *tmplock, int tmplocksz,
		volatile char **xtmplock,
		int retries, int flags, struct __lockargs *args)
{
	struct stat	st, st1;
	char		pidbuf[40];
	pid_t		pid = 0;
	int		sleeptime = 0;
	int		statfailed = 0;
	int		fd;
	int		i, e, pidlen;
	int		dontsleep = 1;
	int		tries = retries + 1;

	/* process optional flags that have arguments */
	if (flags & __L_INTERVAL) {
		sleeptime = args->interval;
	}

	/* decide which PID to write to the lockfile */
	if (flags & L_PID)
		pid = getpid();
	if (flags & L_PPID) {
		pid = getppid();
		if (pid == 1) {
			/* orphaned */
			return L_ORPHANED;
		}
	}
	pidlen = snprintf(pidbuf, sizeof(pidbuf), "%d\n", pid);
	if (pidlen > sizeof(pidbuf) - 1) {
		errno = EOVERFLOW;
		return L_ERROR;
	}

	/* create temporary lockfile */
	if ((i = lockfilename(lockfile, tmplock, tmplocksz)) != 0)
		return i;
	if (xtmplock)
		*xtmplock = tmplock;
	fd = open(tmplock, O_WRONLY|O_CREAT|O_EXCL|O_CLOEXEC, 0644);
	if (fd < 0) {
		/* permission denied? perhaps try suid helper */
#if defined(LIB) && defined(MAILGROUP)
		if (errno == EACCES && is_maillock(lockfile))
			return run_helper("-l", lockfile, retries, flags);
#endif
		return L_TMPLOCK;
	}
	i = write(fd, pidbuf, pidlen);
	e = errno;

	if (close(fd) != 0) {
		e = errno;
		i = -1;
	}
	if (i != pidlen) {
		unlink(tmplock);
		tmplock[0] = 0;
		errno = i < 0 ? e : EAGAIN;
		return L_TMPWRITE;
	}

	/*
	 *	Now try to link the temporary lock to the lock.
	 */
	for (i = 0; i < tries && tries > 0; i++) {
		if (!dontsleep) {
			if (!(flags & __L_INTERVAL))
				sleeptime += 5;

			if (sleeptime > 60) sleeptime = 60;
#ifdef LIB
			sleep(sleeptime);
#else
			if ((e = check_sleep(sleeptime, flags)) != 0) {
				unlink(tmplock);
				tmplock[0] = 0;
				return e;
			}
#endif
		}
		dontsleep = 0;


		/*
		 *	Now lock by linking the tempfile to the lock.
		 *
		 *	KLUDGE: some people say the return code of
		 *	link() over NFS can't be trusted.
		 *	EXTRA FIX: the value of the nlink field
		 *	can't be trusted (may be cached).
		 */
		(void)!link(tmplock, lockfile);

		if (lstat(tmplock, &st1) < 0) {
			tmplock[0] = 0;
			return L_ERROR; /* Can't happen */
		}

		if (lstat(lockfile, &st) < 0) {
			if (statfailed++ > 5) {
				/*
				 *	Normally, this can't happen; either
				 *	another process holds the lockfile or
				 *	we do. So if this error pops up
				 *	repeatedly, just exit...
				 */
				e = errno;
				(void)unlink(tmplock);
				tmplock[0] = 0;
				errno = e;
				return L_MAXTRYS;
			}
			continue;
		}

		/*
		 *	See if we got the lock.
		 */
		if (st.st_rdev == st1.st_rdev &&
		    st.st_ino  == st1.st_ino) {
			(void)unlink(tmplock);
			tmplock[0] = 0;
			return L_SUCCESS;
		}
		statfailed = 0;

		/*
		 *	If there is a lockfile and it is invalid,
		 *	remove the lockfile.
		 */
		if (lockfile_check(lockfile, flags) == -1) {
			if (unlink(lockfile) < 0 && errno != ENOENT) {
				/*
				 *	we failed to unlink the stale
				 *	lockfile, give up.
				 */
				return L_RMSTALE;
			}
			dontsleep = 1;
			/*
			 *	If the lockfile was invalid, then the first
			 *	try wasn't valid either - make sure we
			 *	try at least once more.
			 */
			if (tries == 1) tries++;
		}

	}
	(void)unlink(tmplock);
	tmplock[0] = 0;
	errno = EAGAIN;
	return L_MAXTRYS;
}

#ifdef LIB
static
#endif
int lockfile_create_set_tmplock(const char *lockfile, volatile char **xtmplock, int retries, int flags, struct __lockargs *args)
{
	char *tmplock;
	int l, r, e;

	l = strlen(lockfile)+TMPLOCKFILENAMESZ+1;
	if ((tmplock = (char *)malloc(l)) == NULL)
		return L_ERROR;
	tmplock[0] = 0;
	r = lockfile_create_save_tmplock(lockfile,
						tmplock, l, xtmplock, retries, flags, args);
	if (xtmplock)
		*xtmplock = NULL;
	e = errno;
	free(tmplock);
	errno = e;
	return r;
}

#ifdef LIB
int lockfile_create(const char *lockfile, int retries, int flags)
{
	/* check against unknown flags */
	if (flags & ~(L_PID|L_PPID)) {
		errno = EINVAL;
		return L_ERROR;
	}
	return lockfile_create_set_tmplock(lockfile, NULL, retries, flags, NULL);
}

#ifdef STATIC
int lockfile_create2(const char *lockfile, int retries,
		int flags, struct __lockargs *args, int args_sz)
{

	#define FLAGS_WITH_ARGS (__L_INTERVAL)
	#define KNOWN_FLAGS (L_PID|L_PPID|__L_INTERVAL)

	/* check if size is the same (version check) */
	if (args != NULL && sizeof(struct __lockargs) != args_sz) {
		errno = EINVAL;
		return L_ERROR;
	}
	/* some flags _must_ have a non-null args */
	if (args == NULL && (flags & FLAGS_WITH_ARGS)) {
		errno = EINVAL;
		return L_ERROR;
	}
	/* check against unknown flags */
	if (flags & ~KNOWN_FLAGS) {
		errno = EINVAL;
		return L_ERROR;
	}
	return lockfile_create_set_tmplock(lockfile, NULL, retries, flags, args);
}
#endif

#endif

/*
 *	See if a valid lockfile is present.
 *	Returns 0 if so, -1 if not.
 */
int lockfile_check(const char *lockfile, int flags)
{
	struct stat	st, st2;
	char		buf[16];
	time_t		now;
	pid_t		pid;
	int		fd, len, r;

	if (stat(lockfile, &st) < 0)
		return -1;

	/*
	 *	Get the contents and mtime of the lockfile.
	 */
	time(&now);
	pid = 0;
	if ((fd = open(lockfile, O_RDONLY)) >= 0) {
		/*
		 *	Try to use 'atime after read' as now, this is
		 *	the time of the filesystem. Should not get
		 *	confused by 'atime' or 'noatime' mount options.
		 */
		len = 0;
		if (fstat(fd, &st) == 0 &&
		    (len = read(fd, buf, sizeof(buf))) >= 0 &&
		    fstat(fd, &st2) == 0 &&
		    st.st_atime != st2.st_atime)
			now = st.st_atime;
		close(fd);
		if (len > 0 && (flags & (L_PID|L_PPID))) {
			buf[len] = 0;
			pid = atoi(buf);
		}
	}

	if (pid > 0) {
		/*
		 *	If we have a pid, see if the process
		 *	owning the lockfile is still alive.
		 */
		r = kill(pid, 0);
		if (r == 0 || errno == EPERM)
			return 0;
		if (r < 0 && errno == ESRCH)
			return -1;
		/* EINVAL - FALLTHRU */
	}

	/*
	 *	Without a pid in the lockfile, the lock
	 *	is valid if it is newer than 5 mins.
	 */

	if (now < st.st_mtime + 300)
		return 0;

	return -1;
}

/*
 *	Remove a lock.
 */
int lockfile_remove(const char *lockfile)
{
	if (unlink(lockfile) < 0) {
#if defined(LIB) && defined(MAILGROUP)
		if (errno == EACCES && is_maillock(lockfile))
			return run_helper("-u", lockfile, 0, 0);
#endif
		return errno == ENOENT ? 0 : -1;
	}
	return 0;
}

/*
 *	Touch a lock.
 */
int lockfile_touch(const char *lockfile)
{
#ifdef HAVE_UTIME
	return utime(lockfile, NULL);
#else
	return utimes(lockfile, NULL);
#endif
}

#ifdef LIB
/*
 *	Lock a mailfile. This looks a lot like the SVR4 function.
 *	Arguments: lusername, retries.
 */
int maillock(const char *name, int retries)
{
	char		*p, *mail;
	char		*newlock;
	int		i, e;
	int             len, newlen;

	if (islocked) return 0;

#ifdef MAXPATHLEN
	if (strlen(name) + sizeof(MAILDIR) + 6 > MAXPATHLEN) {
		errno = ENAMETOOLONG;
		return L_NAMELEN;
	}
#endif

	/*
	 *	If $MAIL is for the same username as "name"
	 *	then use $MAIL instead.
	 */

	len = strlen(name)+strlen(MAILDIR)+6;
	mlockfile = (char *)malloc(len);
	if (!mlockfile)
		return L_ERROR;
	sprintf(mlockfile, "%s%s.lock", MAILDIR, name);
	if ((mail = getenv("MAIL")) != NULL) {
		if ((p = strrchr(mail, '/')) != NULL)
			p++;
		else
			p = mail;
		if (strcmp(p, name) == 0) {
			newlen = strlen(mail)+6;
#ifdef MAXPATHLEN
			if (newlen > MAXPATHLEN) {
				errno = ENAMETOOLONG;
				return L_NAMELEN;
			}
#endif
			if (newlen > len) {
				newlock = (char *)realloc (mlockfile, newlen);
				if (newlock == NULL) {
					e = errno;
					free (mlockfile);
					mlockfile = NULL;
					errno = e;
					return L_ERROR;
				}
				mlockfile = newlock;
			}
			sprintf(mlockfile, "%s.lock", mail);
		}
	}
	i = lockfile_create(mlockfile, retries, 0);
	if (i == 0) islocked = 1;

	return i;
}

void mailunlock(void)
{
	if (!islocked) return;
	lockfile_remove(mlockfile);
	free (mlockfile);
	islocked = 0;
}

void touchlock(void)
{
	lockfile_touch(mlockfile);
}
#endif

